File: 071.Simplify Path/Simplify Path.cpp
class Solution {
public:
    static bool isSlash(char c)
    {
        return (c == '/');
    }
    static bool notSlash(char c)
    {
        return !isSlash(c);
    }
    vector<string> split(string str)
    {
        vector<string> ret;
        string::iterator it = str.begin();
        while(it != str.end())
        {
            it = find_if(it, str.end(), notSlash);
            string::iterator it2 = find_if(it, str.end(), isSlash);
            if(it != str.end())
                ret.push_back(string(it, it2));
            it = it2;
        }
        return ret;
    }
    string simplifyPath(string path) {
        vector<string> v = split(path);
        stack<string> stk;
        string ret = "";
        for(int i = 0; i < v.size(); i ++)
        {
            if(v[i] == ".")
                ;
            else if(v[i] == "..")
            {
                if(!stk.empty())
                    stk.pop();
            }
            else
                stk.push(v[i]);
        }
        while(!stk.empty())
        {
            ret = "/" + stk.top() + ret;
            stk.pop();
        }
        if(ret == "")
            return "/";
        return ret;
    }
};

File: 071.Simplify Path/Simplify Path.py
# 题目大意

# 简化linux路径。

# 解题方法

# 看到这种来来回回，增增删删的题，一般都想到用栈。

# 我们把字符串按照/分割之后就得到了每个文件的目录，然后判断路径是添加还是向上层进行返回。这个题很简单了。


class Solution(object):
    def simplifyPath(self, path):
        """
        :type path: str
        :rtype: str
        """
        stack = list()
        dirs = path.split('/')
        for dir in dirs:
            if not dir or dir == '.':
                continue
            if dir == '..':
                if stack:
                    stack.pop()
            else:                
                stack.append(dir)
        return '/' + '/'.join(stack)
